include/AudioPluginInterface.h

```diff
@@ -0,0 +1,418 @@
+/*
+ * AudioPluginInterface.h - Interface for all audio plugins
+ *
+ * Copyright (c) 2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_AUDIO_PLUGIN_INTERFACE_H
+#define LMMS_AUDIO_PLUGIN_INTERFACE_H
+
+#include "AudioData.h"
+#include "AudioPluginBuffer.h"
+#include "Effect.h"
+#include "Instrument.h"
+#include "InstrumentTrack.h"
+#include "PluginPinConnector.h"
+#include "SampleFrame.h"
+
+namespace lmms
+{
+
+enum class ProcessStatus
+{
+	//! Unconditionally continue processing
+	Continue,
+
+	//! Calculate the RMS out sum and call `checkGate` to determine whether to stop processing
+	ContinueIfNotQuiet,
+
+	//! Do not continue processing
+	Sleep
+};
+
+//! Compile time customizations for `AudioPluginInterface` to meet the needs of a plugin implementation
+struct PluginConfig
+{
+	//! The audio data layout used by the plugin
+	AudioDataLayout layout;
+
+	//! The number of plugin input channels, or `DynamicChannelCount` if unknown at compile time
+	int inputs = DynamicChannelCount;
+
+	//! The number of plugin output channels, or `DynamicChannelCount` if unknown at compile time
+	int outputs = DynamicChannelCount;
+
+	//! In-place processing - true (always in-place) or false (dynamic, customizable in audio buffer impl)
+	bool inplace = false;
+
+	//! If true, plugin implementation will provide an `AudioPluginBufferInterfaceProvider`
+	bool customBuffer = false;
+};
+
+class NotePlayHandle;
+
+namespace detail
+{
+
+//! Provides the correct `processImpl` interface for instruments or effects to implement
+template<class ChildT, typename BufferT, typename ConstBufferT, bool inplace, bool customBuffer>
```

Comment:
A general comment: The whole template logic here is well done and readable for a programmer who is used to templates. However, it seems to me that traditionally, LMMS was always coded with a "keep the code easy and simple" principle, so it might be that in the future, some users might complain that they cannot read the base classes of their plugins anymore. However, this principle might be as deprecated as many others, e.g. the "all Instruments shall be 250x250". Also, while the templates themselves might be hard to read, their use is straightforward (similar to STL code).

So, for me, headers like theses are perfectly fine. However, feel free to ask other developers for more feedback about adding more complex template logic into our "Qt-playground-codebase".

Reply:
Yeah, there may be some opportunities to simplify it further but I'd have to give it some thought

Reply:
I would only invest time for that if there are actual complaints from users/reviewers. Feel free to close this comment.